panic: runtime error: slice bounds out of range

```
panic: runtime error: slice bounds out of range

goroutine 83 [running]:
github.com/liftbridge-io/liftbridge/server/commitlog.(*Index).WriteAt(0xc42001f020, 0xc4202fc000, 0x3000, 0x4bf4, 0xa01c7c, 0x0)
        /Users/tylertreat/Go/src/github.com/liftbridge-io/liftbridge/server/commitlog/index.go:154 +0xf3
github.com/liftbridge-io/liftbridge/server/commitlog.(*Index).WriteEntries(0xc42001f020, 0xc421bba000, 0x400, 0x400, 0x400, 0x12400)
        /Users/tylertreat/Go/src/github.com/liftbridge-io/liftbridge/server/commitlog/index.go:108 +0x1b1
github.com/liftbridge-io/liftbridge/server/commitlog.(*Segment).WriteMessageSet(0xc4202d05a0, 0xc421a6e000, 0x12400, 0x1cb03, 0xc421bba000, 0x400, 0x400, 0xc42057ffc0, 0xc42016e1c0)
        /Users/tylertreat/Go/src/github.com/liftbridge-io/liftbridge/server/commitlog/segment.go:111 +0xb3
github.com/liftbridge-io/liftbridge/server/commitlog.(*CommitLog).append(0xc4200ae240, 0xc4202d05a0, 0xc421a6e000, 0x12400, 0x1cb03, 0xc421bba000, 0x400, 0x400, 0xc421bba000, 0x400, ...)
        /Users/tylertreat/Go/src/github.com/liftbridge-io/liftbridge/server/commitlog/commitlog.go:207 +0x74
github.com/liftbridge-io/liftbridge/server/commitlog.(*CommitLog).Append(0xc4200ae240, 0xc420436000, 0x400, 0x400, 0xc421462000, 0x400, 0x400, 0x0, 0x0)
        /Users/tylertreat/Go/src/github.com/liftbridge-io/liftbridge/server/commitlog/commitlog.go:186 +0x1a5
github.com/liftbridge-io/liftbridge/server.(*stream).messageProcessingLoop(0xc4200eed10, 0xc42001f080, 0xc42008aa80)
        /Users/tylertreat/Go/src/github.com/liftbridge-io/liftbridge/server/stream.go:465 +0x2a1
github.com/liftbridge-io/liftbridge/server.(*stream).becomeLeader.func1()
        /Users/tylertreat/Go/src/github.com/liftbridge-io/liftbridge/server/stream.go:243 +0x3f
github.com/liftbridge-io/liftbridge/server.(*Server).startGoroutine.func1(0xc42017e5c0, 0xc4201ca0b0)
        /Users/tylertreat/Go/src/github.com/liftbridge-io/liftbridge/server/server.go:632 +0x27
created by github.com/liftbridge-io/liftbridge/server.(*Server).startGoroutine
        /Users/tylertreat/Go/src/github.com/liftbridge-io/liftbridge/server/server.go:631 +0x99
```

--- a/server/commitlog/index.go
+++ b/server/commitlog/index.go
@@ -29,6 +29,7 @@ type Index struct {
 	options
 	mmap     gommap.MMap
 	file     *os.File
+	size     int64
 	mu       sync.RWMutex
 	position int64
 }
@@ -86,9 +87,9 @@ func NewIndex(opts options) (idx *Index, err error) {
 	fi, err := idx.file.Stat()
 	if err != nil {
 		return nil, errors.Wrap(err, "stat file failed")
-	} else {
-		idx.position = fi.Size()
 	}
+	idx.position = fi.Size()
+	idx.size = fi.Size()
 
 	idx.mmap, err = gommap.Map(idx.file.Fd(), gommap.PROT_READ|gommap.PROT_WRITE, gommap.MAP_SHARED)
 	if err != nil {
@@ -151,6 +152,24 @@ func (idx *Index) ReadAt(p []byte, offset int64) (n int, err error) {
 func (idx *Index) WriteAt(p []byte, offset int64) (n int) {
 	idx.mu.Lock()
 	defer idx.mu.Unlock()
+
+	// Check if we need to expand the index file.
+	if offset >= idx.size {
+		// Expand the index file.
+		newSize := roundDown(idx.size+idx.bytes, entryWidth)
+		err := idx.file.Truncate(newSize)
+		if err != nil {
+			panic(errors.Wrap(err, "failed to expand index file"))
+		}
+		idx.size = newSize
+
+		// Re-mmap the index.
+		idx.mmap, err = gommap.Map(idx.file.Fd(), gommap.PROT_READ|gommap.PROT_WRITE, gommap.MAP_SHARED)
+		if err != nil {
+			panic(errors.Wrap(err, "failed to mmap expanded index file"))
+		}
+	}
+
 	return copy(idx.mmap[offset:], p)
 }
 

